internal/procedures/server/start.go

```diff
@@ -47,15 +47,49 @@ func Start(initiator common.ServerInitiator) (err error) {
 		go startCourse(course)
 	}
 
-	// Cleanup any temp dirs.
-	defer util.RemoveRecordedTempDirs()
+	return nil
+}
 
-	err = server.RunServer(initiator)
-	if err != nil {
-		return fmt.Errorf("Error during server startup sequence: '%w'.", err)
+func CleanupAndStopServer() (err error) {
+	if serverInstance == nil {
+		return nil
 	}
 
-	log.Debug("Server closed.")
+	err = errors.Join(err, db.Close())
+	err = errors.Join(err, util.RemoveRecordedTempDirs())
+
+	serverInstance.apiServer.StopServer()
+
+	serverInstance.apiServer = nil
+	serverInstance = nil
+
+	return err
+}
+
+func RunAndBlockServer(initiator common.ServerInitiator, skipSetup bool) (err error) {
+	// Run inside a func so defers will run before the function returns.
+	func() {
+		defer func() {
+			err = errors.Join(err, CleanupAndStopServer())
+		}()
+
+		serverInstance = &ProcedureServer{
+			apiServer: server.NewAPIServer(),
+		}
+
+		if !skipSetup {
+			err = setupServer(initiator)
+			if err != nil {
+				err = fmt.Errorf("Failed to setup the server: '%w'.", err)
+				return
+			}
+		}
+
+		err = serverInstance.apiServer.RunServer(initiator)
```

Comment:
You need a comment here noting that serverInstance may be nil when this call completes.